### PR TITLE
Add support for podman and streaming images from read-only registries with some other artifact storage.

### PR DESCRIPTION
**Pull request by iain-macdonald**
This PR does a few things:

1. Adds support for podman by porting the [store/fs.go ](https://github.com/containerd/stargz-snapshotter/blob/main/store/fs.go)file (and dependencies) from the stargz-snapshotter into the soci-snapshotter and hooking it up to a new binary that was added, the soci-store.
2. Adds a hook that makes the snapshotter attempt to retrieve artifacts from the local filesystem before the local OCI store and remote container registry. This is used in our set up to support streaming images from registries that we can't write to, the executor will fetch the artifacts elsewhere and store them in the right place locally.
3. Adds a hook that looks up the image-to-index mapping from the local filesystem before the remote container registry. This is, again, used to support streaming images from registries that we can't write to. Note that the mapping is from the hash of the image's config which changes when the image contents change, not its tag which may not change.

This is pretty similar to the version of the soci-snapshotter we have running in production now, though I rebased it to the [awslabs/soci-snapshotter](https://github.com/awslabs/soci-snapshotter) head and cleaned it up a bit.

### CODE REVIEW

**Comment by iain-macdonald** [comment]
FYI diff is much more reasonable if you compare the `store/` files against their [stargz-snapshotter](https://github.com/containerd/stargz-snapshotter) counterparts.

File: fs/artifact_fetcher.go [hunk 1]
```diff
@@ -102,25 +103,50 @@ func (f *artifactFetcher) constructRef(desc ocispec.Descriptor) string {
 	return fmt.Sprintf("%s@%s", f.refspec.Locator, desc.Digest.String())
 }
 
+type nopCloser struct {
+	io.Reader
```
Comment (vadimberezniker) [inline fs/artifact_fetcher.go]: golang has `io.NopCloser`

File: fs/artifact_fetcher.go [hunk 2]
```diff
@@ -102,25 +103,50 @@ func (f *artifactFetcher) constructRef(desc ocispec.Descriptor) string {
...
+}
+
 // Fetches the artifact identified by the descriptor.
 // It first checks the local store for the artifact.
 // If not found, if constructs the ref and fetches it from remote.
 func (f *artifactFetcher) Fetch(ctx context.Context, desc ocispec.Descriptor) (io.ReadCloser, bool, error) {
+	// Try to read the requested artifact from the local filesystem first.
+	// This is faster and lets us bypass all of the container registry interaction when available.
+	// TODO(iain): pass this directory as an argument or parameter.
+	localFilename := "/var/lib/soci-snapshotter-grpc/content/blobs/sha256/" + desc.Digest.Encoded()
```
Comment (vadimberezniker) [inline fs/artifact_fetcher.go]: Let's at least make this directory a constant if we can't make it a parameter yet.

**Review by vadimberezniker — APPROVED** [review]

File: fs/artifact_fetcher.go (hunk 1 above)
Comment (iain-macdonald) [reply fs/artifact_fetcher.go]: Done

File: fs/artifact_fetcher.go (hunk 2 above)
Comment (iain-macdonald) [reply fs/artifact_fetcher.go]: Done

**Pull request merged**